File: notes/mcp/note-service.js
import path from "node:path";
import fsp from "node:fs/promises";
import { legacyNotesDataFile, notesDataFile } from "./config.js";

const MAX_TEXT_LENGTH = 100 * 1024;
const MAX_TAGS = 20;
const MAX_TAG_LENGTH = 60;

export class NoteService {
  constructor() {
    this.notes = [];
    this.nextId = 1;
  }

  async load() {
    const file = notesDataFile();
    await this.#migrateLegacyIfNeeded(file);
    try {
      const raw = await fsp.readFile(file, "utf8");
      const data = JSON.parse(raw);
      if (Array.isArray(data.notes)) {
        this.notes = data.notes.map((note) => this._migrate(note));
        const maxId = Math.max(0, ...this.notes.map((n) => Number(n.id) || 0));
        this.nextId = maxId + 1;
      }
    } catch (error) {
      if (error.code !== "ENOENT") {
        process.stderr.write(`[notes-mcp] failed to load notes: ${error.message}\n`);
      }
    }
  }

  async #migrateLegacyIfNeeded(targetFile) {
    // Explicit test/override paths must stay empty until the service creates them.
    if (process.env.NUSASHELL_NOTES_DATA_FILE) return;
    const legacy = legacyNotesDataFile();
    if (path.resolve(legacy) === path.resolve(targetFile)) return;
    try {
      await fsp.access(targetFile);
      return;
    } catch {
      // Target missing — try one-time copy from plugin-adjacent legacy file.
    }
    try {
      const raw = await fsp.readFile(legacy, "utf8");
      await fsp.mkdir(path.dirname(targetFile), { recursive: true });
      await fsp.writeFile(targetFile, raw, "utf8");
      process.stderr.write(`[notes-mcp] migrated notes from legacy path to ${targetFile}\n`);
    } catch (error) {
      if (error.code !== "ENOENT") {
        process.stderr.write(`[notes-mcp] legacy notes migration skipped: ${error.message}\n`);
      }
    }
  }

  async save() {
    const file = notesDataFile();
    await fsp.mkdir(path.dirname(file), { recursive: true });
    await fsp.writeFile(file, JSON.stringify({ notes: this.notes }, null, 2), "utf8");
  }

  _migrate(note) {
    return {
      id: Number(note.id) || 0,
      text: String(note.text ?? ""),
      tags: Array.isArray(note.tags) ? note.tags.map(String) : [],
      createdAt: note.createdAt || new Date().toISOString(),
      updatedAt: note.updatedAt || note.createdAt || new Date().toISOString(),
    };
  }

  _validateText(text) {
    if (typeof text !== "string") throw new Error("text must be a string");
    if (text.length === 0) throw new Error("text must not be empty");
    if (text.length > MAX_TEXT_LENGTH) {
      throw new Error(`text too long (${text.length} chars, max ${MAX_TEXT_LENGTH})`);
    }
  }

  _validateTags(tags) {
    if (!Array.isArray(tags)) throw new Error("tags must be an array");
    if (tags.length > MAX_TAGS) throw new Error(`too many tags (max ${MAX_TAGS})`);
    for (const tag of tags) {
      if (typeof tag !== "string" || tag.length === 0) {
        throw new Error("each tag must be a non-empty string");
      }
      if (tag.length > MAX_TAG_LENGTH) {
        throw new Error(`tag too long (max ${MAX_TAG_LENGTH} chars): ${tag.slice(0, 20)}...`);
      }
    }
  }

  create(text, tags = []) {
    this._validateText(text);
    this._validateTags(tags);
    const now = new Date().toISOString();
    const note = {
      id: this.nextId++,
      text,
      tags,
      createdAt: now,
      updatedAt: now,
    };
    this.notes.push(note);
    return note;
  }

  list(tag, sort = "updated") {
    let notes = tag ? this.notes.filter((n) => n.tags.includes(tag)) : [...this.notes];
    const sortKey = sort === "created" ? "createdAt" : "updatedAt";
    notes.sort((a, b) => {
      const aVal = a[sortKey] || a.createdAt;
      const bVal = b[sortKey] || b.createdAt;
      return new Date(bVal).getTime() - new Date(aVal).getTime();
    });
    return notes;
  }

  get(id) {
    const note = this.notes.find((n) => n.id === Number(id));
    if (!note) throw new Error(`Note not found: ${id}`);
    return note;
  }

  update(id, updates) {
    const note = this.get(id);
    if (updates.text !== undefined) {
      this._validateText(updates.text);
      note.text = updates.text;
    }
    if (updates.tags !== undefined) {
      this._validateTags(updates.tags);
      note.tags = updates.tags;
    }
    note.updatedAt = new Date().toISOString();
    return note;
  }

  delete(id) {
    const idx = this.notes.findIndex((n) => n.id === Number(id));
    if (idx === -1) throw new Error(`Note not found: ${id}`);
    const [deleted] = this.notes.splice(idx, 1);
    return deleted;
  }

  search(query) {
    let regex;
    try {
      regex = new RegExp(query, "i");
    } catch {
      regex = new RegExp(query.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), "i");
    }
    return this.notes.filter((n) => regex.test(n.text));
  }

  allTags() {
    const counts = {};
    for (const note of this.notes) {
      for (const tag of note.tags) {
        counts[tag] = (counts[tag] || 0) + 1;
      }
    }
    return Object.entries(counts)
      .map(([tag, count]) => ({ tag, count }))
      .sort((a, b) => b.count - a.count);
  }
}

File: notes/mcp/note_service.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

const (
	maxTextLength = 100 * 1024
	maxTags       = 20
	maxTagLength  = 60
)

// Note is a single persisted note.
type Note struct {
	ID        int      `json:"id"`
	Text      string   `json:"text"`
	Tags      []string `json:"tags"`
	CreatedAt string   `json:"createdAt"`
	UpdatedAt string   `json:"updatedAt"`
}

// NoteService manages in-memory notes with JSON file persistence.
type NoteService struct {
	dataFile string
	notes    []Note
	nextID   int
}

// NewNoteService creates a service backed by dataFile.
func NewNoteService(dataFile string) *NoteService {
	return &NoteService{dataFile: dataFile, nextID: 1}
}

// Load reads notes from the data file. Missing file is not an error.
func (s *NoteService) Load() error {
	raw, err := os.ReadFile(s.dataFile)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("read notes: %w", err)
	}

	var data struct {
		Notes []Note `json:"notes"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return fmt.Errorf("parse notes: %w", err)
	}

	s.notes = make([]Note, 0, len(data.Notes))
	maxID := 0
	for _, n := range data.Notes {
		n = migrateNote(n)
		if n.ID > maxID {
			maxID = n.ID
		}
		s.notes = append(s.notes, n)
	}
	s.nextID = maxID + 1
	return nil
}

// Save writes notes to the data file, creating parent dirs as needed.
func (s *NoteService) Save() error {
	dir := filepath.Dir(s.dataFile)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("create data dir: %w", err)
	}

	data := struct {
		Notes []Note `json:"notes"`
	}{Notes: s.notes}

	raw, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal notes: %w", err)
	}
	if err := os.WriteFile(s.dataFile, raw, 0o644); err != nil {
		return fmt.Errorf("write notes: %w", err)
	}
	return nil
}

func migrateNote(n Note) Note {
	if n.Tags == nil {
		n.Tags = []string{}
	}
	now := time.Now().UTC().Format(time.RFC3339)
	if n.CreatedAt == "" {
		n.CreatedAt = now
	}
	if n.UpdatedAt == "" {
		n.UpdatedAt = n.CreatedAt
	}
	return n
}

func validateText(text string) error {
	if text == "" {
		return fmt.Errorf("text must not be empty")
	}
	if len(text) > maxTextLength {
		return fmt.Errorf("text too long (%d chars, max %d)", len(text), maxTextLength)
	}
	return nil
}

func validateTags(tags []string) error {
	if len(tags) > maxTags {
		return fmt.Errorf("too many tags (max %d)", maxTags)
	}
	for _, tag := range tags {
		if tag == "" {
			return fmt.Errorf("each tag must be a non-empty string")
		}
		if len(tag) > maxTagLength {
			return fmt.Errorf("tag too long (max %d chars): %s...", maxTagLength, tag[:min(20, len(tag))])
		}
	}
	return nil
}

// Create adds a new note and returns it.
func (s *NoteService) Create(text string, tags []string) (Note, error) {
	if err := validateText(text); err != nil {
		return Note{}, err
	}
	if err := validateTags(tags); err != nil {
		return Note{}, err
	}
	now := time.Now().UTC().Format(time.RFC3339)
	n := Note{
		ID:        s.nextID,
		Text:      text,
		Tags:      tags,
		CreatedAt: now,
		UpdatedAt: now,
	}
	s.nextID++
	s.notes = append(s.notes, n)
	return n, nil
}

// List returns notes optionally filtered by tag, sorted by updatedAt (default)
// or createdAt.
func (s *NoteService) List(tag, sortKey string) []Note {
	var out []Note
	for _, n := range s.notes {
		if tag != "" {
			found := false
			for _, t := range n.Tags {
				if t == tag {
					found = true
					break
				}
			}
			if !found {
				continue
			}
		}
		out = append(out, n)
	}

	key := "updatedAt"
	if sortKey == "created" {
		key = "createdAt"
	}

	sort.SliceStable(out, func(i, j int) bool {
		a, b := out[i], out[j]
		aVal := fieldByPath(a, key)
		bVal := fieldByPath(b, key)
		return aVal > bVal // newest first
	})

	return out
}

func fieldByPath(n Note, key string) string {
	switch key {
	case "createdAt":
		return n.CreatedAt
	default:
		return n.UpdatedAt
	}
}

// Get returns a note by ID.
func (s *NoteService) Get(id int) (Note, error) {
	for _, n := range s.notes {
		if n.ID == id {
			return n, nil
		}
	}
	return Note{}, fmt.Errorf("Note not found: %d", id)
}

// NoteUpdate holds optional fields for Update.
type NoteUpdate struct {
	Text *string
	Tags *[]string
}

// Update modifies a note's text and/or tags.
func (s *NoteService) Update(id int, updates NoteUpdate) (Note, error) {
	idx := -1
	for i, n := range s.notes {
		if n.ID == id {
			idx = i
			break
		}
	}
	if idx == -1 {
		return Note{}, fmt.Errorf("Note not found: %d", id)
	}

	n := s.notes[idx]
	if updates.Text != nil {
		if err := validateText(*updates.Text); err != nil {
			return Note{}, err
		}
		n.Text = *updates.Text
	}
	if updates.Tags != nil {
		if err := validateTags(*updates.Tags); err != nil {
			return Note{}, err
		}
		n.Tags = *updates.Tags
	}
	n.UpdatedAt = time.Now().UTC().Format(time.RFC3339)
	s.notes[idx] = n
	return n, nil
}

// Delete removes a note by ID and returns the deleted note.
func (s *NoteService) Delete(id int) (Note, error) {
	for i, n := range s.notes {
		if n.ID == id {
			s.notes = append(s.notes[:i], s.notes[i+1:]...)
			return n, nil
		}
	}
	return Note{}, fmt.Errorf("Note not found: %d", id)
}

// Search returns notes whose text matches a case-insensitive regex pattern.
// Invalid patterns are escaped and retried.
func (s *NoteService) Search(query string) []Note {
	re, err := regexp.Compile("(?i)" + query)
	if err != nil {
		escaped := regexp.QuoteMeta(query)
		re, err = regexp.Compile("(?i)" + escaped)
		if err != nil {
			return nil
		}
	}

	var out []Note
	for _, n := range s.notes {
		if re.MatchString(n.Text) {
			out = append(out, n)
		}
	}
	return out
}

// AllTags returns tag counts sorted by frequency (descending).
func (s *NoteService) AllTags() []TagCount {
	counts := map[string]int{}
	for _, n := range s.notes {
		for _, t := range n.Tags {
			counts[t]++
		}
	}
	out := make([]TagCount, 0, len(counts))
	for tag, count := range counts {
		out = append(out, TagCount{Tag: tag, Count: count})
	}
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].Count != out[j].Count {
			return out[i].Count > out[j].Count
		}
		return strings.Compare(out[i].Tag, out[j].Tag) < 0
	})
	return out
}

// TagCount is a tag name with its usage count.
type TagCount struct {
	Tag   string `json:"tag"`
	Count int    `json:"count"`
}

// Total returns the number of stored notes.
func (s *NoteService) Total() int {
	return len(s.notes)
}
